pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpsSegmentFetcher.java

```diff
@@ -0,0 +1,169 @@
+    } catch (Exception e) {
+      Utils.rethrowException(e);
+    }
+  }
+
+  private TrustManager[] setupTrustManagers(Configuration configs)
+      throws CertificateException, KeyStoreException, IOException, NoSuchAlgorithmException {
+    // This is the cert authority that validates server's cert, so we need to put it in our
+    // trustStore.
+    String serverCACertFile = configs.getString(CONFIG_OF_SERVER_CA_CERT);
+    if (serverCACertFile == null) {
+      LOGGER.warn("Https server CA Certificate file not confugured (" + CONFIG_OF_SERVER_CA_CERT + "). All servers will be trusted!");
+      TrustManager[] trustAllCerts = new TrustManager[] {
+          new X509TrustManager() {
+            @Override
+            public void checkClientTrusted(X509Certificate[] x509Certificates, String s) throws CertificateException {
+            }
+
+            @Override
+            public void checkServerTrusted(X509Certificate[] x509Certificates, String s) throws CertificateException {
+            }
+
+            @Override
+            public X509Certificate[] getAcceptedIssuers() {
+              return new X509Certificate[0];
+            }
+          }
+      };
+      return trustAllCerts;
+    }
+    FileInputStream is = new FileInputStream(new File(serverCACertFile));
+    CertificateFactory certificateFactory = CertificateFactory.getInstance(CERTIFICATE_TYPE);
+    X509Certificate cert = (X509Certificate) certificateFactory.generateCertificate(is);
+
+    String serverKey = "https-server";
+    KeyStore trustStore = KeyStore.getInstance(KeyStore.getDefaultType());
+    trustStore.load(null);
+    trustStore.setCertificateEntry(serverKey, cert);
+
+    TrustManagerFactory tmf = TrustManagerFactory.getInstance(CERTIFICATE_TYPE);
+    tmf.init(trustStore);
+    return tmf.getTrustManagers();
+  }
+
+  private KeyManager[] setupKeyManagers(Configuration configs) {
+    try {
+      KeyStore keyStore = KeyStore.getInstance(KEYSTORE_TYPE);
+      String keyStoreFile = configs.getString(CONFIG_OF_CLIENT_PKCS12_FILE);
+      String keyStorePassword = configs.getString(CONFIG_OF_CLIENT_PKCS12_PASSWORD);
+      if (keyStoreFile == null || keyStorePassword == null) {
+        LOGGER.info("Either keystore file name (" + CONFIG_OF_CLIENT_PKCS12_FILE + ") or " +
+            "keystore password (" + CONFIG_OF_CLIENT_PKCS12_PASSWORD + ") is not configured"
+            + ". Client will not present certificate to the server");
+        return null;
+      }
+      keyStore.load(new FileInputStream(new File(keyStoreFile)), keyStorePassword.toCharArray());
+      KeyManagerFactory kmf = KeyManagerFactory.getInstance(KEYMANAGER_FACTORY_ALGORITHM);
+      kmf.init(keyStore, keyStorePassword.toCharArray());
+      return kmf.getKeyManagers();
+    } catch (Exception e) {
+      Utils.rethrowException(e);
+    }
+    return null;
+  }
+
+  @Override
+  public void fetchSegmentToLocal(String uri, File tempFile) throws Exception {
+    URL url = new URL(uri);
```

Comment:
do we want to implement retries here like we do in httpsegmentfetcher?

Reply:
No, this class will do the same as HttpSegmentFetcher, returning either a permanent or temporary exception, and let the layer above retry.

Reply:
Yes, good point, I will add retries.